`|=` operator on `bool` results in invalid Rust being generated

I tried running c2rust on [BRRtools](https://github.com/Optiroc/BRRtools) as an example of a small but nontrivial C project. However the generated Rust fails to compile. The input C in question is:

https://github.com/Optiroc/BRRtools/blob/9f6c4708150120284102f9899632d7f5f0b0638a/src/brr_encoder.c#L596-L598

I minified this example to the following C code (`x` can be either `bool` or `int`):

```c
#include <stdbool.h>

// both bool and int reproduce the bug
extern bool x;

int main() {
	bool initial_block = false;
	initial_block |= x;
}
```

Running the latest c2rust 030df665741bb9d1b29fe41cdbfd7223a67c5995 produces the following Rust code:

```rust
#![allow(dead_code, mutable_transmutes, non_camel_case_types, non_snake_case,
         non_upper_case_globals, unused_assignments, unused_mut)]
#![register_tool(c2rust)]
#![feature(main, register_tool)]
use ::c2rust_out::*;
extern "C" {
    // both bool and int reproduce the bug
    #[no_mangle]
    static mut x: bool;
}
unsafe fn main_0() -> libc::c_int {
    let mut initial_block: bool = 0 as libc::c_int != 0;
    initial_block = (initial_block as libc::c_int | x as libc::c_int) as bool;
    return 0;
}
#[main]
pub fn main() { unsafe { ::std::process::exit(main_0() as i32) } }
```

which produces an error when compiling on both `cargo +nightly-2019-12-05` and stable:

```
error[E0054]: cannot cast as `bool`
  --> src/brr_encoder.rs:13:21
   |
13 |     initial_block = (initial_block as libc::c_int | x as libc::c_int) as bool;
   |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: compare with zero instead: `(initial_block as libc::c_int | x as libc::c_int) != 0`
```

I've uploaded at repo at https://github.com/nyanpasu64/c2rust-bug. Steps to reproduce:

```
c2rust transpile --binary brr_encoder --overwrite-existing compile_commands.json
cargo build
```

I'm guessing this is also the same issue:
```C
	_Bool q = 1;
	q += 1;
```

```Rust
    let mut q: bool = 1 as libc::c_int != 0;
    q = (q as libc::c_int + 1 as libc::c_int) as bool;
```

```
error[E0054]: cannot cast as `bool`
   --> src/binary.rs:245:9
    |
245 |     q = (q as libc::c_int + 1 as libc::c_int) as bool;
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: compare with zero instead: `(q as libc::c_int + 1 as libc::c_int) != 0`

```

--- a/c2rust-transpile/src/translator/mod.rs
+++ b/c2rust-transpile/src/translator/mod.rs
@@ -4149,7 +4149,7 @@ impl<'c> Translation<'c> {
         }
     }
 
-    fn convert_cast(
+    pub fn convert_cast(
         &self,
         ctx: ExprContext,
         source_cty: CQualTypeId,

--- a/c2rust-transpile/src/translator/operators.rs
+++ b/c2rust-transpile/src/translator/operators.rs
@@ -206,6 +206,7 @@ impl<'c> Translation<'c> {
         read: Box<Expr>,
         write: Box<Expr>,
         rhs: Box<Expr>,
+        initial_lhs_type_id: CQualTypeId,
         compute_lhs_type_id: CQualTypeId,
         compute_res_type_id: CQualTypeId,
         lhs_type_id: CQualTypeId,
@@ -220,46 +221,33 @@ impl<'c> Translation<'c> {
                 rhs,
             )))
         } else {
-            let compute_lhs_resolved_ty = &self.ast_context.resolve_type(compute_lhs_type_id.ctype);
-            let lhs_type = self.convert_type(compute_lhs_type_id.ctype)?;
-
-            // We can't simply as-cast into a non primitive like f128
-            let lhs = if compute_lhs_resolved_ty.kind == CTypeKind::LongDouble {
-                self.use_crate(ExternCrate::F128);
-
-                let fn_path = mk().abs_path_expr(vec!["f128", "f128", "from"]);
-                let args = vec![read];
-
-                mk().call_expr(fn_path, args)
-            } else {
-                mk().cast_expr(read, lhs_type)
-            };
-            let ty = self.convert_type(compute_res_type_id.ctype)?;
-            let mut val = self.convert_binary_operator(
+            let lhs = self.convert_cast(
                 ctx,
-                bin_op,
-                ty,
-                compute_res_type_id.ctype,
+                initial_lhs_type_id,
                 compute_lhs_type_id,
-                rhs_type_id,
-                lhs,
-                rhs,
+                WithStmts::new_val(read.clone()),
+                None,
+                None,
                 None,
             )?;
 
-            let resolve_lhs_kind = &self.ast_context.resolve_type(lhs_type_id.ctype).kind;
+            let ty = self.convert_type(compute_res_type_id.ctype)?;
+            let val = lhs.and_then(|lhs| {
+                self.convert_binary_operator(
+                    ctx,
+                    bin_op,
+                    ty,
+                    compute_res_type_id.ctype,
+                    compute_lhs_type_id,
+                    rhs_type_id,
+                    lhs,
+                    rhs,
+                    None,
+                )
+            })?;
 
-            val = if let &CTypeKind::Enum(enum_id) = resolve_lhs_kind {
-                val.result_map(|val| {
-                    self.convert_cast_to_enum(ctx, lhs_type_id.ctype, enum_id, None, val)
-                })?
-            } else if compute_lhs_resolved_ty.kind == CTypeKind::LongDouble {
-                // We can't as-cast from a non primitive like f128 back to the result_type
-                self.f128_cast_to(val, resolve_lhs_kind)?
-            } else {
-                let result_type = self.convert_type(lhs_type_id.ctype)?;
-                val.map(|val| mk().cast_expr(val, result_type))
-            };
+            let val =
+                self.convert_cast(ctx, compute_res_type_id, lhs_type_id, val, None, None, None)?;
 
             Ok(val.map(|val| mk().assign_expr(write.clone(), val)))
         }
@@ -452,29 +440,26 @@ impl<'c> Translation<'c> {
 
                         // Anything volatile needs to be desugared into explicit reads and writes
                         op if is_volatile || is_unsigned_arith => {
+                            // Cast the lhs to the compute lhs type, do the compute, and then
+                            // cast the compute result to the final lhs type.
+
                             let op = op
                                 .underlying_assignment()
                                 .expect("Cannot convert non-assignment operator");
 
-                            let val = if expr_or_comp_type_id.ctype == initial_lhs_type_id.ctype {
+                            let lhs = self.convert_cast(
+                                ctx,
+                                initial_lhs_type_id,
+                                expr_or_comp_type_id,
+                                WithStmts::new_val(read.clone()),
+                                None,
+                                None,
+                                None,
+                            )?;
+
+                            let ty = self.convert_type(result_type_id.ctype)?;
+                            let val = lhs.and_then(|lhs|
                                 self.convert_binary_operator(
-                                    ctx,
-                                    op,
-                                    ty,
-                                    expr_type_id.ctype,
-                                    initial_lhs_type_id,
-                                    rhs_type_id,
-                                    read.clone(),
-                                    rhs,
-                                    None,
-                                )?
-                            } else {
-                                let lhs_type =
-                                    self.convert_type(compute_lhs_type_id.unwrap().ctype)?;
-                                let write_type = self.convert_type(expr_type_id.ctype)?;
-                                let lhs = mk().cast_expr(read.clone(), lhs_type);
-                                let ty = self.convert_type(result_type_id.ctype)?;
-                                let mut val = self.convert_binary_operator(
                                     ctx,
                                     op,
                                     ty,
@@ -484,25 +469,18 @@ impl<'c> Translation<'c> {
                                     lhs,
                                     rhs,
                                     None,
-                                )?;
+                                )
+                            )?;
 
-                                let expr_resolved_ty_kind = &self.ast_context.resolve_type(expr_type_id.ctype).kind;
-
-                                val = if let &CTypeKind::Enum(enum_id) = expr_resolved_ty_kind {
-                                    val.result_map(|val| self.convert_cast_to_enum(
-                                        ctx,
-                                        expr_type_id.ctype,
-                                        enum_id,
-                                        None,
-                                        val,
-                                    ))?
-                                } else {
-                                    let expr_type = self.convert_type(expr_type_id.ctype)?;
-                                    val.map(|val| mk().cast_expr(val, expr_type))
-                                };
-
-                                val.map(|val| mk().cast_expr(val, write_type))
-                            };
+                            let val = self.convert_cast(
+                                ctx,
+                                result_type_id,
+                                expr_type_id,
+                                val,
+                                None,
+                                None,
+                                None,
+                            )?;
 
                             #[allow(clippy::let_and_return /* , reason = "block is large, so variable name helps" */)]
                             let write = if is_volatile {
@@ -554,6 +532,7 @@ impl<'c> Translation<'c> {
                                 read.clone(),
                                 write,
                                 rhs,
+                                initial_lhs_type_id,
                                 compute_lhs_type_id.unwrap(),
                                 compute_res_type_id.unwrap(),
                                 expr_type_id,

--- a/c2rust-transpile/tests/snapshots.rs
+++ b/c2rust-transpile/tests/snapshots.rs
@@ -309,10 +309,7 @@ fn test_atomics() {
 
 #[test]
 fn test_bool() {
-    transpile("bool.c")
-        .expect_compile_error_edition_2021(true)
-        .expect_compile_error_edition_2024(true)
-        .run();
+    transpile("bool.c").run();
 }
 
 #[test]

--- a/c2rust-transpile/tests/snapshots/bool.c
+++ b/c2rust-transpile/tests/snapshots/bool.c
@@ -27,7 +27,6 @@ void test_bool() {
     MACRO_INT1 ? 2 : 3;
     MACRO_IDENT1 ? 2 : 3;
 
-    // Doesn't compile currently.
-    // See https://github.com/immunant/c2rust/issues/340.
+    // Tests https://github.com/immunant/c2rust/issues/340
     int0 |= int1;
 }

--- a/c2rust-transpile/tests/snapshots/exprs.c
+++ b/c2rust-transpile/tests/snapshots/exprs.c
@@ -34,6 +34,16 @@ void unary_with_side_effect(){
     arr[side_effect()]--;
 }
 
+void unsigned_compound_desugaring(void) {
+    enum E { EA };
+
+    int i = 0;
+    unsigned int u = 0;
+    enum E e = EA;
+    e += u;
+    i += u;
+}
+
 void compound_literal(){
     /// https://github.com/immunant/c2rust/issues/1234
     int i = (enum {A, B, C}){1};

